src/Spice86.Core/Emulator/Memory/ExtendedMemoryManager.cs

```diff
@@ -0,0 +1,485 @@
+        int handle = (ushort)_machine.Cpu.State.DX;
+
+        if (!this.handles.TryGetValue(handle, out int lockCount)) {
+            _machine.Cpu.State.AX = 0; // Didn't work.
+            _machine.Cpu.State.BL = 0xA2; // Invalid handle.
+            return;
+        }
+
+        this.handles[handle] = lockCount + 1;
+
+        _ = this.TryGetBlock(handle, out XmsBlock block);
+        uint fullAddress = XmsBaseAddress + block.Offset;
+
+        _machine.Cpu.State.AX = 1; // Success.
+        _machine.Cpu.State.DX = (ushort)(fullAddress >> 16);
+        _machine.Cpu.State.BX = (ushort)(fullAddress & 0xFFFFu);
+    }
+    /// <summary>
+    /// Unlocks a block of memory.
+    /// </summary>
+    private void UnlockBlock() {
+        int handle = (ushort)_machine.Cpu.State.DX;
+
+        if (!this.handles.TryGetValue(handle, out int lockCount)) {
+            _machine.Cpu.State.AX = 0; // Didn't work.
+            _machine.Cpu.State.BL = 0xA2; // Invalid handle.
+            return;
+        }
+
+        if (lockCount < 1) {
+            _machine.Cpu.State.AX = 0;
+            _machine.Cpu.State.BL = 0xAA; // Handle is not locked.
+            return;
+        }
+
+        this.handles[handle] = lockCount - 1;
+
+        _machine.Cpu.State.AX = 1; // Success.
+    }
+    /// <summary>
+    /// Returns information about an XMS handle.
+    /// </summary>
+    private void GetHandleInformation() {
+        int handle = (ushort)_machine.Cpu.State.DX;
+
+        if (!this.handles.TryGetValue(handle, out int lockCount)) {
+            _machine.Cpu.State.AX = 0; // Didn't work.
+            _machine.Cpu.State.BL = 0xA2; // Invalid handle.
+            return;
+        }
+
+        _machine.Cpu.State.BH = (byte)lockCount;
+        _machine.Cpu.State.BL = (byte)(MaxHandles - this.handles.Count);
+
+        if (!this.TryGetBlock(handle, out XmsBlock block)) {
+            _machine.Cpu.State.DX = 0;
+        } else {
+            _machine.Cpu.State.DX = (ushort)((block).Length / 1024u);
+        }
+
+        _machine.Cpu.State.AX = 1; // Success.
+    }
+    /// <summary>
+    /// Copies a block of memory.
+    /// </summary>
+    private void MoveMemoryBlock() {
+        bool a20State = this._machine.Memory.EnableA20;
+        this._machine.Memory.EnableA20 = true;
+
+        XmsMoveData moveData;
+        unsafe {
+            moveData = *(XmsMoveData*)this._machine.Memory.GetPointer(this._machine.Cpu.State.DS, this._machine.Cpu.State.SI);
+        }
+
+        IntPtr srcPtr = IntPtr.Zero;
+        IntPtr destPtr = IntPtr.Zero;
+
+        if (moveData.SourceHandle == 0) {
+            SegmentedAddress srcAddress = moveData.SourceAddress;
+            srcPtr = this._machine.Memory.GetPointer(srcAddress.Segment, srcAddress.Offset);
+        } else {
+            if (this.TryGetBlock(moveData.SourceHandle, out XmsBlock srcBlock)) {
+                srcPtr = this._machine.Memory.GetPointer((int)(XmsBaseAddress + (srcBlock).Offset + moveData.SourceOffset));
```

Comment:
I think code would be much simpler if instead of storing the block data in emulated memory (which cannot be accessed past 1M in real mode anyway) we stored them as byte array in a proper c# datastructure